File: src/commands/entity_analysis.rs
//! Unified entity analysis command.
//!
//! Consolidates entity lifecycle tracking, creation analysis, preprocessing,
//! gap detection, and timeline analysis into a single powerful command with
//! intelligent auto-preprocessing to eliminate multi-step workflows.
//!
//! # Usage
//!
//! ```bash
//! # Churn analysis (auto-preprocesses entity mappings)
//! vault-audit entity-analysis churn logs/day1.log logs/day2.log
//! vault-audit entity-analysis churn logs/*.log --baseline entities.json
//!
//! # Creation analysis by auth path
//! vault-audit entity-analysis creation logs/*.log
//! vault-audit entity-analysis creation logs/*.log --export creation_data.json
//!
//! # Extract entity mappings (preprocessing)
//! vault-audit entity-analysis preprocess logs/*.log --output mappings.json
//! vault-audit entity-analysis preprocess logs/*.log --format csv
//!
//! # Detect activity gaps for entities
//! vault-audit entity-analysis gaps logs/*.log --window-seconds 300
//!
//! # Individual entity timeline
//! vault-audit entity-analysis timeline logs/*.log --entity-id abc-123
//! ```
//!
//! **Key Improvement**: Auto-preprocessing eliminates the need for separate
//! preprocessing steps. Entity mappings are built in-memory automatically when
//! needed by churn or creation analysis.
//!
//! # Subcommands
//!
//! ## churn
//! Multi-day entity lifecycle tracking with ephemeral pattern detection.
//! Automatically preprocesses entity mappings unless `--no-auto-preprocess` is specified.
//!
//! ## creation
//! Analyzes when entities were first created, grouped by authentication path.
//! Shows new entity onboarding patterns and growth trends.
//!
//! ## preprocess
//! Extracts entity-to-display-name mappings from audit logs for external use.
//! Outputs JSON or CSV format for integration with other tools.
//!
//! ## gaps
//! Detects entities with suspicious activity gaps (potential compromised credentials
//! or entities that should have been cleaned up).
//!
//! ## timeline
//! Shows chronological activity for a specific entity ID, useful for debugging
//! or investigating specific identity issues.

use anyhow::Result;
use std::fs::File;
use std::io::Write;

/// Helper to write entity map to temp JSON file for commands that expect file paths
fn write_temp_entity_map(
    entity_map: &std::collections::HashMap<
        String,
        crate::commands::preprocess_entities::EntityMapping,
    >,
) -> Result<String> {
    let temp_path = format!(".vault-audit-autopreprocess-{}.json", std::process::id());

    let file = File::create(&temp_path)?;
    let mut writer = std::io::BufWriter::new(file);
    let json = serde_json::to_string_pretty(&entity_map)?;
    writer.write_all(json.as_bytes())?;
    writer.flush()?;

    Ok(temp_path)
}

/// Run churn analysis subcommand
pub fn run_churn(
    log_files: &[String],
    entity_map: Option<&String>,
    baseline: Option<&String>,
    output: Option<&String>,
    format: Option<&String>,
    auto_preprocess: bool,
) -> Result<()> {
    // Auto-preprocessing: build entity map in-memory and write to temp file
    let temp_map_file = if auto_preprocess && entity_map.is_none() {
        eprintln!("Auto-preprocessing: Building entity mappings in-memory...\n");
        let map = crate::commands::preprocess_entities::build_entity_map(log_files)?;
        let temp_path = write_temp_entity_map(&map)?;
        eprintln!("Entity mappings ready\n");
        Some(temp_path)
    } else {
        None
    };

    // Use provided map or auto-generated temp map
    let map_to_use = entity_map
        .map(std::string::String::as_str)
        .or(temp_map_file.as_deref());

    // Delegate to existing entity_churn implementation
    let result = crate::commands::entity_churn::run(
        log_files,
        map_to_use,
        baseline.map(std::string::String::as_str),
        output.map(std::string::String::as_str),
        format.map(std::string::String::as_str),
    );

    // Cleanup temp file
    if let Some(temp) = temp_map_file {
        let _ = std::fs::remove_file(temp);
    }

    result
}

/// Run creation analysis subcommand
pub fn run_creation(
    log_files: &[String],
    entity_map: Option<&String>,
    output: Option<&String>,
    auto_preprocess: bool,
) -> Result<()> {
    // Auto-preprocessing: build entity map in-memory and write to temp file
    let temp_map_file = if auto_preprocess && entity_map.is_none() {
        eprintln!("Auto-preprocessing: Building entity mappings in-memory...\n");
        let map = crate::commands::preprocess_entities::build_entity_map(log_files)?;
        let temp_path = write_temp_entity_map(&map)?;
        eprintln!("Entity mappings ready\n");
        Some(temp_path)
    } else {
        None
    };

    // Use provided map or auto-generated temp map
    let map_to_use = entity_map
        .map(std::string::String::as_str)
        .or(temp_map_file.as_deref());

    // Delegate to existing entity_creation implementation
    let result = crate::commands::entity_creation::run(
        log_files,
        map_to_use,
        output.map(std::string::String::as_str),
    );

    // Cleanup temp file
    if let Some(temp) = temp_map_file {
        let _ = std::fs::remove_file(temp);
    }

    result
}

/// Run preprocess subcommand
pub fn run_preprocess(log_files: &[String], output: &str, format: &str) -> Result<()> {
    // Delegate to existing preprocess_entities implementation
    crate::commands::preprocess_entities::run(log_files, output, format)
}

/// Run gaps detection subcommand
pub fn run_gaps(log_files: &[String], window_seconds: u64) -> Result<()> {
    // Delegate to existing entity_gaps implementation
    crate::commands::entity_gaps::run(log_files, window_seconds)
}

/// Run timeline subcommand
pub fn run_timeline(
    log_files: &[String],
    entity_id: &str,
    display_name: Option<&String>,
) -> Result<()> {
    // Delegate to existing entity_timeline implementation
    crate::commands::entity_timeline::run(log_files, entity_id, display_name)
}

File: internal/commands/entity_analysis.go
// Package commands provides subcommands for vault audit analysis.
package commands

import (
	"encoding/json"
	"fmt"
	"os"
	"time"

	"github.com/trenner1/hashicorp-vault-audit-analysis/internal/audit"
	"github.com/trenner1/hashicorp-vault-audit-analysis/internal/processor"
	"github.com/trenner1/hashicorp-vault-audit-analysis/internal/utils"
)

// EntityMapping represents entity-to-display-name mappings extracted from audit logs.
type EntityMapping struct {
	DisplayName   string `json:"display_name"`
	MountPath     string `json:"mount_path"`
	MountAccessor string `json:"mount_accessor"`
	Username      string `json:"username,omitempty"`
	LoginCount    int    `json:"login_count"`
	FirstSeen     string `json:"first_seen"`
	LastSeen      string `json:"last_seen"`
}

// buildEntityMap extracts entity mappings from audit log files.
func buildEntityMap(logFiles []string) (map[string]EntityMapping, error) {
	type state struct {
		entityMap map[string]EntityMapping
	}

	result, _, err := processor.RunFiles(
		processor.DefaultConfig(),
		logFiles,
		func() state {
			return state{entityMap: make(map[string]EntityMapping)}
		},
		func(entry *audit.AuditEntry, s *state) {
			// Look for login events in auth paths
			path := entry.Path()
			if !entry.PathStartsWith("auth/") || !pathContains(path, "/login") {
				return
			}

			if entry.Auth == nil || entry.Auth.EntityID == nil {
				return
			}

			entityID := *entry.Auth.EntityID
			if entityID == "" {
				return
			}

			displayName := ""
			if entry.Auth.DisplayName != nil {
				displayName = *entry.Auth.DisplayName
			}
			if displayName == "" {
				return
			}

			// Extract mount path (e.g., "auth/github/login" -> "auth/github")
			mountPath := trimSuffixes(path, "/login", "/"+displayName)

			mountAccessor := ""
			if entry.Auth.Accessor != nil {
				mountAccessor = *entry.Auth.Accessor
			}

			username := ""
			if entry.MetadataString("username") != "" {
				username = entry.MetadataString("username")
			}

			if mapping, exists := s.entityMap[entityID]; exists {
				mapping.LoginCount++
				mapping.LastSeen = entry.Time
				s.entityMap[entityID] = mapping
			} else {
				s.entityMap[entityID] = EntityMapping{
					DisplayName:   displayName,
					MountPath:     mountPath,
					MountAccessor: mountAccessor,
					Username:      username,
					LoginCount:    1,
					FirstSeen:     entry.Time,
					LastSeen:      entry.Time,
				}
			}
		},
		func(a, b state) state {
			// Merge entity maps
			for id, mapping := range b.entityMap {
				if existing, ok := a.entityMap[id]; ok {
					existing.LoginCount += mapping.LoginCount
					existing.LastSeen = mapping.LastSeen
					a.entityMap[id] = existing
				} else {
					a.entityMap[id] = mapping
				}
			}
			return a
		},
	)

	if err != nil {
		return nil, err
	}

	return result.entityMap, nil
}

// writeTempEntityMap writes entity mappings to a temporary JSON file.
func writeTempEntityMap(entityMap map[string]EntityMapping) (string, error) {
	tempPath := fmt.Sprintf(".vault-audit-autopreprocess-%d.json", os.Getpid())

	data, err := json.MarshalIndent(entityMap, "", "  ")
	if err != nil {
		return "", err
	}

	if err := os.WriteFile(tempPath, data, 0600); err != nil {
		return "", err
	}

	return tempPath, nil
}

// RunChurn performs entity churn analysis with auto-preprocessing.
func RunChurn(logFiles []string, entityMap, baseline, output, format *string, autoPreprocess bool) error {
	var tempMapFile string

	// Auto-preprocessing: build entity map in-memory and write to temp file
	if autoPreprocess && (entityMap == nil || *entityMap == "") {
		fmt.Fprintln(os.Stderr, "Auto-preprocessing: Building entity mappings in-memory...")
		entityMapData, err := buildEntityMap(logFiles)
		if err != nil {
			return err
		}
		tempPath, err := writeTempEntityMap(entityMapData)
		if err != nil {
			return err
		}
		tempMapFile = tempPath
		fmt.Fprintln(os.Stderr, "Entity mappings ready")
	}

	// Use provided map or auto-generated temp map
	var mapToUse *string
	if entityMap != nil && *entityMap != "" {
		mapToUse = entityMap
	} else if tempMapFile != "" {
		mapToUse = &tempMapFile
	}

	// Delegate to entity_churn implementation
	err := RunEntityChurn(logFiles, mapToUse, baseline, output, format)

	// Cleanup temp file
	if tempMapFile != "" {
		_ = os.Remove(tempMapFile)
	}

	return err
}

// RunCreation performs entity creation analysis with auto-preprocessing.
func RunCreation(logFiles []string, entityMap, output *string, autoPreprocess bool) error {
	var tempMapFile string

	// Auto-preprocessing: build entity map in-memory and write to temp file
	if autoPreprocess && (entityMap == nil || *entityMap == "") {
		fmt.Fprintln(os.Stderr, "Auto-preprocessing: Building entity mappings in-memory...")
		entityMapData, err := buildEntityMap(logFiles)
		if err != nil {
			return err
		}
		tempPath, err := writeTempEntityMap(entityMapData)
		if err != nil {
			return err
		}
		tempMapFile = tempPath
		fmt.Fprintln(os.Stderr, "Entity mappings ready")
	}

	// Use provided map or auto-generated temp map
	var mapToUse *string
	if entityMap != nil && *entityMap != "" {
		mapToUse = entityMap
	} else if tempMapFile != "" {
		mapToUse = &tempMapFile
	}

	// Delegate to entity_creation implementation
	err := RunEntityCreation(logFiles, mapToUse, output)

	// Cleanup temp file
	if tempMapFile != "" {
		_ = os.Remove(tempMapFile)
	}

	return err
}

// RunPreprocess extracts and exports entity mappings.
func RunPreprocess(logFiles []string, output, format string) error {
	return RunEntityPreprocess(logFiles, output, format)
}

// RunGaps detects activity gaps for entities.
func RunGaps(logFiles []string, windowSeconds uint64) error {
	return RunEntityGaps(logFiles, windowSeconds)
}

// RunTimeline shows chronological activity for a specific entity.
func RunTimeline(logFiles []string, entityID string, displayName *string) error {
	return RunEntityTimeline(logFiles, entityID, displayName)
}

// Helper functions

func pathContains(path, substr string) bool {
	for i := 0; i <= len(path)-len(substr); i++ {
		if path[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}

func trimSuffixes(s string, suffixes ...string) string {
	for _, suffix := range suffixes {
		if len(s) >= len(suffix) && s[len(s)-len(suffix):] == suffix {
			return s[:len(s)-len(suffix)]
		}
	}
	return s
}

// loadEntityMappingsFromFile loads entity mappings from JSON or CSV file.
func loadEntityMappingsFromFile(path string) (map[string]EntityMapping, error) {
	// Try JSON first
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	var mappings map[string]EntityMapping
	if err := json.Unmarshal(data, &mappings); err == nil {
		return mappings, nil
	}

	// If JSON fails, could also try CSV, but for now just fail
	return nil, fmt.Errorf("failed to parse entity mappings from %s", path)
}

// loadBaselineEntities loads entity baseline from JSON file.
type baselineEntity struct {
	EntityID      string `json:"entity_id"`
	EntityName    string `json:"entity_name"`
	EntityCreated string `json:"entity_created"`
	AliasName     string `json:"alias_name"`
	MountPath     string `json:"mount_path"`
	MountType     string `json:"mount_type"`
	MountAccessor string `json:"mount_accessor"`
}

func (b *baselineEntity) getName() string {
	if b.EntityName != "" {
		return b.EntityName
	}
	return b.AliasName
}

func loadBaselineEntities(path string) (map[string]baselineEntity, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	var entities []baselineEntity
	if err := json.Unmarshal(data, &entities); err != nil {
		return nil, err
	}

	result := make(map[string]baselineEntity)
	for _, e := range entities {
		if _, exists := result[e.EntityID]; !exists {
			result[e.EntityID] = e
		}
	}
	return result, nil
}

// parseTime parses RFC3339 timestamp strings.
func parseTime(ts string) time.Time {
	t, err := utils.ParseTimestamp(ts)
	if err != nil {
		return time.Time{}
	}
	return t
}
